debug: new pages for List and listIndexed

stefanhengl opened the pull request:
This adds new debug pages which were previously only exposed as CLI debug commands.

/debug/list
/debug/indexed?indexed=<TRUE, false>

<img width="1103" alt="Screenshot 2022-05-19 at 14 35 56" src="https://user-images.githubusercontent.com/26413131/169294669-f2d00f41-bfaf-4c52-aa8c-443a81097e06.png">

Other changes:

- We only show the PProf link if PProf is enabled
- Add link to "Queue" on the debug landing page

stefanhengl committed:
debug: new pages for List, Indexed

This adds new debug pages

/debug/list
/debug/indexed?indexed=<FALSE, true>

Other changes

- / -> /debug/reindex
- /debug is now the landing page containing links to all other pages
- We only show the PProf button if PProf is enabled

stefanhengl commented:
just realized I cannot use queue.get as I wanted. I will update the PR and set it to draft for now.

stefanhengl committed:
add queue.get and lock the queue properly

stefanhengl committed:
minor: add title

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -614,6 +617,65 @@ func (s *Server) handleHTTPRoot(w http.ResponseWriter, r *http.Request) {
+	indexedStr := r.URL.Query().Get("indexed")
+	if b, err := strconv.ParseBool(indexedStr); err != nil {
+		withIndexed = b
+	}
```
if you are ignoring the error (which seems fine) then this works since it will return false on failure:

```suggestion
	withIndexed, _ := strconv.ParseBool(r.URL.Query().Get("indexed"))
```

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -614,6 +617,65 @@ func (s *Server) handleHTTPRoot(w http.ResponseWriter, r *http.Request) {
+	}
+
+	var indexed []uint32
+	if withIndexed {
+		indexed = listIndexed(s.IndexDir)
+	}
```
should the default be withIndexed=true? zoekt normally cares about including this info. However, it is useful to not include it.

stefanhengl replied:
agree. Default is now "true".

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -614,6 +617,65 @@ func (s *Server) handleHTTPRoot(w http.ResponseWriter, r *http.Request) {
+		indexed = listIndexed(s.IndexDir)
+	}
+
+	repos, err := s.Sourcegraph.List(context.Background(), indexed)
```
```suggestion
	repos, err := s.Sourcegraph.List(r.Context(), indexed)
```

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -614,6 +617,65 @@ func (s *Server) handleHTTPRoot(w http.ResponseWriter, r *http.Request) {
+	}
+
+	s.queue.mu.Lock()
+	defer s.queue.mu.Unlock()
```
maybe me overengineering here, but this will hold the lock while the user is doing the request. For slow connections (eg mine) this means the queue can be blocked for a while. This data shouldn't be that huge, instead I'd write to a buffer with lock, then write it all in one go.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -614,6 +617,65 @@ func (s *Server) handleHTTPRoot(w http.ResponseWriter, r *http.Request) {
+
+		_, err := fmt.Fprintf(w, "%d\t%s\n", id, name)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
```
if you fail to write the use has hung up, so this line doesn't do anything. Normally one will debug log or something.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -861,8 +923,14 @@ func startServer(conf rootConfig) error {
+				{Href: "debug/list?indexed=true", Text: "Assigned (all)", Description: "includes repositories which this instance temporarily holds during re-balancing"},
+				{Href: "debug/list?indexed=false", Text: "Assigned (this instance)"},
+				{Href: "debug/queue", Text: "Queue"},
+				{Href: "debug/reindex", Text: "Re-index"},
```
can you test these relative links work when proxied via the sourcegraph admin instrumentation proxy? I worry it may do something surprising about the final link it creates.

keegancsmith reviewed (approved):
LGTM. some minor changes inline, but approving to unblock.

stefanhengl committed:
Update cmd/zoekt-sourcegraph-indexserver/main.go

Co-authored-by: Keegan Carruthers-Smith <keegan.csmith@gmail.com>

stefanhengl committed:
write to local buffer while holding the lock

stefanhengl committed:
fix relative links on site admin proxy view

The site admin view in Sourcegraph requires the services to handle a
request at their respective root "/". Hence we move the re-index view
back to the root.

At the same time we fix a bug in production. Clicking on re-index
when comming from the Sourcegraph site admin view sent the form to the
absolute instead of the relative root.

stefanhengl committed:
use tabwritter

stefanhengl committed:
default to /debug/list?indexed=true

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -614,6 +623,114 @@ func (s *Server) handleHTTPRoot(w http.ResponseWriter, r *http.Request) {
+			debug.Printf("handleDebugList: %s\n", err.Error())
+		}
+	}
+	defer s.queue.mu.Unlock()
```
this is a defer. I think you meant to unlock right away :P
```suggestion
	s.queue.mu.Unlock()
```

stefanhengl committed:
Update cmd/zoekt-sourcegraph-indexserver/main.go

Co-authored-by: Keegan Carruthers-Smith <keegan.csmith@gmail.com>

stefanhengl committed:
rename <style>: profile-name -> debug-page
